### ISSUE TITLE
VisualInstance(s) are spamming the VisualServer with useless updates on low FPS

### ISSUE THREAD

**Issue by m4nu3lf** [godotengine/godot#55089]
### Godot version

45088780be124acc8ab226d3e5cee808bb626980

### System information

OS: Arch Linux, CPU: Ryzen 5, GUP: RX 480

### Issue description

VisualInstance(s) notify the VisualServer through the command queue (if MT is enabled) that the transform has changed whenever the transform notifications are flushed, which happens every physics frame. This means that if the rendering framerate is lower than the physics frame rate, the VisualServer will receive many useless notifications making things worse.

Here is the perf flame graph for the current implementation:

![Screenshot_20211118_132548](https://user-images.githubusercontent.com/5478365/142425866-3532abcb-84b9-4e52-8d0c-00b33be5a924.png)

Here is the time spent notifying the visual server of the current implementation:
![Screenshot_20211118_133019](https://user-images.githubusercontent.com/5478365/142426046-ddba8da2-2574-4389-8079-33b273069297.png)

Here is the flame graph for an optimized implementation that uses an internal flag and only notifies the VisualServer on the "internal process" notification:
![Screenshot_20211118_132651](https://user-images.githubusercontent.com/5478365/142426262-83bf6925-1dd5-42d7-8eeb-27d335bb82fa.png)

Here is the time spent notifying the visual server with the same optimization as before:
![Screenshot_20211118_133127](https://user-images.githubusercontent.com/5478365/142426418-0fefb732-b37d-41bd-a386-3633f840a837.png)

That's around 2% speedup on the main thread, but considering that the benchmark I'm using is dominated by scripting and physics, and there aren't that many visual instances, it could be much more in some other cases.

### Steps to reproduce

The best way to reproduce this is to add rigid bodies to the scene with child mesh instances while profiling.
Add enough of them, and the rendering FPS will drop below the physics FPS.
Profiling with perf will show the issue.

**Comment by m4nu3lf** [comment]
Just profiled on a game level. Although not perfect as it's hard to replicate the same conditions across multiple runs, the time spent on transform notifications for the VisualServer seems to go from ~3.4% to 0.36%

**Comment by lawnjelly** [comment]
Just to throw a spanner in the works, you won't be able to do this shortly (in the naive way at least) because we're adding fixed timestep interpolation, which needs both the previous and most recent transforms from the last 2 physics ticks.

So if you think about the typical situation:
```
physics 60tps, with 30fps rendering you will still need to send 2/2 transforms.
physics 60tps, with 20fps you can potentially save 1/3 of the transforms.
physics 60tps, with 15fps you can potentially save 2/4 transforms.
```

So unless you are e.g. running at a higher physics tick rate, there's a danger that this is a catch 22 - it only starts getting at all useful once performance has tanked anyway, and maybe not many would be playing such a low performing game. It's a good observation, but also weighing up the real world benefit versus any added complexity in the code and potential for bugs is important. And presumably it's only appearing in the profile when there are a large number of objects, in most cases it may not make any noticeable difference.

An alternative way of approaching this when interpolation is in, would be to send two ticks of transforms in one 'packet', if this ends up being more efficient. But again I don't know if the added complexity would be worth the small gains, especially in 3.x. You could try to persuade reduz for 4.x maybe.

**Comment by Calinou** [comment]
High physics FPS can be useful for simulation driving games where physics FPS is often set at 200 or more, since it's needed for accurate simulation.

**Comment by lawnjelly** [comment]
> High physics FPS can be useful for simulation driving games where physics FPS is often set at 200 or more, since it's needed for accurate simulation.

Yup sure, I don't mean to dismiss the approach, it is very valid observation. :+1:

**Comment by m4nu3lf** [comment]
@lawnjelly Correct me if I'm wrong, but the physics step interpolation only affects cases when the rendering frequency is higher than the physics update one.
If this is the case, we should be able to treat the last two updates that go to the VisualServer like the previous two coming from the physics step.
If this is not the case, it means that the rendering fps is lower than the physics one, and interpolation would be useless anyway.

**Comment by lawnjelly** [comment]
> @lawnjelly Correct me if I'm wrong, but the physics step interpolation only affects cases when the rendering frequency is higher than the physics update one. If this is the case, we should be able to treat the last two updates that go to the VisualServer like the previous two coming from the physics step. If this is not the case, it means that the rendering fps is lower than the physics one, and interpolation would be useless anyway.

Interpolation needs the last two physics ticks in all cases, even when rendering slower than physics.
